examples/stabilizer_code.py

```diff
@@ -0,0 +1,192 @@
+            [
+                np.zeros((k, r), dtype=np.int8),
+                E.T,
+                np.eye(k, dtype=np.int8),
+                np.mod(E.T @ C1.T + C2.T, 2),
+                np.zeros((k, n - k - r), np.int8),
+                np.zeros((k, k), np.int8),
+            ],
+            axis=1,
+        )
+
+        Z = np.concatenate(
+            [
+                np.zeros((k, n), dtype=np.int8),
+                A2.T,
+                np.zeros((k, n - k - r), dtype=np.int8),
+                np.eye(k, dtype=np.int8),
+            ],
+            axis=1,
+        )
+
+        self.n = n
+        self.k = k
+        self.r = r
+
+        def _BuildByCode(mat):
+            out = []
+            n = mat.shape[1] // 2
+            for i in range(mat.shape[0]):
+                ps = ''
+                for j in range(n):
+                    if mat[i, j] == 0 and mat[i, j + n] == 0:
+                        ps += 'I'
+                    elif mat[i, j] == 1 and mat[i, j + n] == 0:
+                        ps += 'X'
+                    elif mat[i, j] == 0 and mat[i, j + n] == 1:
+                        ps += 'Z'
+                    else:
+                        ps += 'Y'
+                out.append(ps)
+            return out
+
+        self.M = _BuildByCode(M)
+        self.X = _BuildByCode(X)
+        self.Z = _BuildByCode(Z)
+        self.syndromes_to_corrections = {}
+
+        for qid in range(self.n):
+            for op in ['X', 'Z']:
+                syndrome = [
+                    1 if self.M[r][qid] == 'I' or self.M[r][qid] == op else -1
+                    for r in range(self.r)
+                ]
+                self.syndromes_to_corrections[tuple(syndrome)] = (op, qid)
+
+    def encode(self, circuit, qubits):
+        # Equation 4.8:
+        for r, x in enumerate(self.X):
+            for j in range(self.r, self.n - self.k):
+                if x[j] == 'X' or x[j] == 'Y':  # coverage: ignore
+                    circuit.append(  # coverage: ignore
+                        cirq.ControlledOperation(  # coverage: ignore
+                            [qubits[self.n - self.k + r]], cirq.X(qubits[j])  # coverage: ignore
+                        )  # coverage: ignore
+                    )  # coverage: ignore
+
+        gate_dict = {'X': cirq.X, 'Y': cirq.Y, 'Z': cirq.Z}
+
+        for r in range(self.r):
+            circuit.append(cirq.H(qubits[r]))
+
+            if self.M[r][r] == 'Y' or self.M[r][r] == 'Z':
+                circuit.append(cirq.S(qubits[r]))
+
+            for n in range(self.n):
+                if n == r:
+                    continue
+                if self.M[r][n] == 'I':
+                    continue
+                op = gate_dict[self.M[r][n]]
+                circuit.append(cirq.ControlledOperation([qubits[r]], op(qubits[n])))
+        # At this stage, the state vector should be equal to equations 3.17 and 3.18.
+
+    def correct(self, circuit, qubits, ancillas):
```

Comment:
As with `encode`, it may be simpler to generate and return a `correction` circuit here, rather than passing in an existing circuit.

Reply:
Done. Looks nicer indeed.